src/mpid/ch4/shm/posix/posix_pre.h

```diff
@@ -32,17 +104,78 @@ typedef struct {
     struct MPIR_Segment *segment_ptr;
     size_t segment_first;
     size_t segment_size;
-} MPIDI_POSIX_request_t;
 
-typedef struct {
-    int dummy;
-} MPIDI_POSIX_comm_t;
+    MPIDI_POSIX_am_request_header_t *req_hdr;
+
+#ifdef POSIX_AM_REQUEST_INLINE
+    MPIDI_POSIX_am_request_header_t req_hdr_buffer;
+#endif                          /* POSIX_AM_REQUEST_INLINE */
+} MPIDI_POSIX_am_request_t;
+
+#define MPIDI_POSIX_EAGER_RECV_INITIALIZE_HOOK(request)\
+do { \
+    (request)->dev.ch4.shm.posix.eager_recv_posted_hook_state = MPIDI_POSIX_EAGER_RECV_POSTED_HOOK_STATE_INITIALIZED; \
+} while (0)
+
+#define MPIDI_POSIX_EAGER_RECV_POSTED_HOOK(request,rank,communicator)\
+do { \
+    if ((request) && ((request)->dev.ch4.shm.posix.eager_recv_posted_hook_state == MPIDI_POSIX_EAGER_RECV_POSTED_HOOK_STATE_INITIALIZED)) { \
+        int grank_ = ((rank) >= 0) ? MPIDI_CH4U_rank_to_lpid((rank), (communicator)) : (rank); \
+        (request)->dev.ch4.shm.posix.eager_recv_posted_hook_state = MPIDI_POSIX_EAGER_RECV_POSTED_HOOK_STATE_REGISTERED; \
+        (request)->dev.ch4.shm.posix.eager_recv_posted_hook_grank = grank_; \
+        MPIDI_POSIX_eager_recv_posted_hook(grank_); \
+    } \
+} while (0)
+
+#define MPIDI_POSIX_EAGER_RECV_COMPLETED_HOOK(request)\
+do { \
+    if ((request)->dev.ch4.shm.posix.eager_recv_posted_hook_state == MPIDI_POSIX_EAGER_RECV_POSTED_HOOK_STATE_REGISTERED) { \
+        MPIDI_POSIX_eager_recv_completed_hook((request)->dev.ch4.shm.posix.eager_recv_posted_hook_grank); \
+    } \
+    (request)->dev.ch4.shm.posix.eager_recv_posted_hook_state = MPIDI_POSIX_EAGER_RECV_POSTED_HOOK_STATE_FINALIZED; \
+} while (0)
```

Comment:
I am not sure why we need these HOOK macros. Are they just used for state tracking ?  Can we disable it when `--enable-fast` is set ?

Reply:
It is a key feature. It is optimization to minimize number of processed unexpected messages. But this key feature doesn't used in simple shm transports like iqueue and fbox.

An advanced shm transport (for example shm transport used in the Intel MPI 2019) must have information about expected or unexpected an available message (request already posted or not yet). If the request was not being posted yet, an advanced shm transport may start process another (expected) message from another rank. Messages from different ranks can be reordered, processing of unexpected message can be retarded.

Reply:
If it is not used in current `fbox` or `iqueue`, we should exclude it in this PR. When we add future transport that requires this feature , we can add the code back.

Reply:
I tend to agree with @minsii here. Until it's needed in the public version of MPICH, it probably shouldn't be here.

Reply:
It does look like the hook functions are necessary however. I guess we just don't need these macros.

Reply:
So it looks like the `MPIDI_POSIX_EAGER_RECV_POSTED_HOOK` (and the function it calls, `MPIDI_POSIX_eager_recv_posted_hook`) _is_ actually used in the fbox transport. I'm not sure we should remove this code after all.

Reply:
It looks like `next_poll_local_rank` is being used in the fbox transport. Can we only keep this part of code in the PR ? The status changes seem unnecessary.

Reply:
I think the state change is necessary in order to make sure that doing the other things are valid. The state change is only used inside these macros, but it still acts as a useful check.

Reply:
I think you're right. The state change might not be necessary. It would be nice to keep the macros though because of the translation from communicator rank to global rank.

I think the state should be removed (and removed from the request structure) and the rest should stay.